app/jobs/regular/notify_tag_change.rb

```diff
@@ -7,9 +7,31 @@ def execute(args)
 
       if post&.topic&.visible?
         post_alerter = PostAlerter.new
-        post_alerter.notify_post_users(post, User.where(id: args[:notified_user_ids]), include_topic_watchers: !post.topic.private_message?, include_category_watchers: false)
+        post_alerter.notify_post_users(post, excluded_users(args), include_topic_watchers: !post.topic.private_message?, include_category_watchers: false)
         post_alerter.notify_first_post_watchers(post, post_alerter.tag_watchers(post.topic))
       end
     end
+
+    private
+
+    def excluded_users(args)
+      return User.where(id: args[:notified_user_ids]) if !args[:diff_tags] || !all_tags_in_hidden_groups?(args)
```

Comment:
Nitpick but since the `if` has two clauses it would probably be better with multiple lines than a one liner.